oeps/oep-0026-arch-realtime-events.rst

```diff
@@ -0,0 +1,403 @@
+
+Use Case: University
+~~~~~~~~~~~~~~~~~~~~
+
+Real-time events also allow the Open edX system to immediately notify
+Universities of their high-interest learning activities, such as unenrollments.
+Note that, in contrast to Enterprises, Universities typically favor Caliper_ as
+their standard of choice.
+
+Use Case: Cross-LMS Integration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Another emerging use case is allowing unification of events across multiple
+LMS's. As a specific example, an organization that uses multiple Open edX
+instances (as multiple options for their learners or for distributing their
+reusable content) may want to accumulate all their LMS events in a single
+`Learning Record Store`_ that is then queried in real-time.
+
+.. _Learning Record Store: https://xapi.com/learning-record-store/
+
+Use Case: External Service
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As we move towards creating a more extensible platform for Open edX, we want to
+provide integration options between the Open edX platform and external services.
+External services that enhance the learning, authoring, or LMS administrative
+experiences will benefit greatly from a real-time eventing framework. Some
+examples include an external social learning service and a commercial learning
+analytics tool. *Theoretically, the other use cases are specific cases of this*
+*use case*.
+
+.. image:: oep-0026/use_cases.png
+
+Technical Specification
+-----------------------
+
+This section provides some of the technical details of the design proposal.
+
+Standards
+=========
+
+In order to enable a plug-n-play ecosystem with various external services
+(adaptive engines or otherwise), we will support both the ADL_'s `Experience
+API (xAPI)`_ standard and the `IMS Global`_'s `Caliper Sensor APIs`_ standard as
+the communication protocol between the Open edX LMS and registered external
+services.
+
+.. _ADL: https://adlnet.gov/
+.. _Experience API (xAPI): https://www.adlnet.gov/research/performance-tracking-analysis/experience-api/
+.. _IMS Global: https://www.imsglobal.org/
+.. _Caliper Sensor APIs: https://www.imsglobal.org/caliper-analytics-v1-public-repos-sensor-apis
+
+xAPI Integration
+~~~~~~~~~~~~~~~~
+
+For details on integrating with xAPI, please see the `xAPI Real-time Events`_
+design document.
+
+.. _xAPI Real-time Events: oep-0026/xapi-realtime-events.rst
+
+Caliper Integration
+~~~~~~~~~~~~~~~~~~~
+
+For details on integrating with Caliper, please see the `Caliper Real-time
+Events`_ design document.
+
+.. _Caliper Real-time Events: oep-0026/caliper-realtime-events.rst
+
+Anonymized User ID
+==================
+
+In order to uniquely identify a user in the Open edX system with a long-term
+and non-revealing identifier, an *Anonymized User ID* will be generated as a 
+SHA-256 hash of the Open edX user_id. The Open edX user_id is the **id** value
```

Comment:
How will we prevent user ID collisions when trying to aggregate events sourced from different Open edX installations?

Reply:
As discussed and captured in the notes: The event itself has the source information, so the source + anonymous-id would be unique.